WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift

```diff
@@ -97,27 +97,83 @@ private extension BlogDashboardService {
 
     func parse(_ entity: BlogDashboardRemoteEntity?, blog: Blog, dotComID: Int) -> [DashboardCardModel] {
         let personalizationService = BlogDashboardPersonalizationService(repository: repository, siteID: dotComID)
-        var cards: [DashboardCardModel] = DashboardCard.allCases.compactMap { card in
-            guard personalizationService.isEnabled(card) else {
+
+        // Map `DashboardCard` instances to `DashboardCardModel`
+        var allCards: [DashboardCardModel] = DashboardCard.allCases.compactMap { card -> DashboardCardModel? in
+            guard card != .dynamic else {
                 return nil
             }
+            return self.dashboardCardModel(
+                from: card,
+                entity: entity,
+                blog: blog,
+                dotComID: dotComID,
+                personalizationService: personalizationService
+            )
+        }
 
-            guard card.shouldShow(
-                for: blog,
-                apiResponse: entity,
-                isJetpack: isJetpack,
-                isDotComAvailable: isDotComAvailable,
-                shouldShowJetpackFeatures: shouldShowJetpackFeatures
-            ) else {
-                return nil
+        // Maps dynamic cards to `DashboardCardModel`.k
+        if let dynamic = entity?.dynamic?.value {
+            let cards = dynamic.compactMap { payload in
+                return self.dashboardDynamicCardModel(for: blog, payload: payload, dotComID: dotComID)
+            }
+            let cardsByOrder = Dictionary(grouping: cards) { $0.payload.order ?? .bottom }
+            let topCards = cardsByOrder[.top, default: []].map { DashboardCardModel.dynamic($0) }
+            let bottomCards = cardsByOrder[.bottom, default: []].map { DashboardCardModel.dynamic($0) }
+
+            // Adds "top" cards at the beginning of the list.
+            allCards = topCards + allCards
+
+            // Adds "bottom" cards at the bottom of the list just before "personalize" card.
+            if allCards.last?.cardType == .personalize {
+                allCards.insert(contentsOf: bottomCards, at: allCards.endIndex - 1)
+            } else {
+                allCards = allCards + bottomCards
             }
+        }
+
+        // Add "empty" card if the list of cards is empty.
+        if allCards.isEmpty || allCards.map(\.cardType) == [.personalize] {
+            let model = DashboardCardModel.normal(.init(cardType: .empty, dotComID: dotComID))
+            allCards.insert(model, at: 0)
+        }
```

Comment:
Would it be possible to unit test this in isolation? Perhaps create a static function if testing it with the current setup requires too much refactor?

Unit testing this kind of logic could pay in dividends .

Reply:
I agree @alpavanoglu, I'll look into unit testing that logic.

Reply:
@alpavanoglu Resolved in c0f54f59c0169d18c0549871b9969dca1d335310